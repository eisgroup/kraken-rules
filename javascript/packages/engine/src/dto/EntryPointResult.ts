/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { FieldEvaluationResult } from "./FieldEvaluationResult";
import { RuleEvaluationResults } from "./RuleEvaluationResults";
import RuleEvaluationResult = RuleEvaluationResults.RuleEvaluationResult;

import { Reducer } from "declarative-js";
import flat = Reducer.flat;
import { ConditionEvaluationResult } from "./ConditionEvaluationResult";

/**
 * Contains evaluation results from specific entry point
 */
export class EntryPointResult {
    private readonly results: Record<string, FieldEvaluationResult>;
    readonly evaluationTimestamp: Date;
    constructor(results: Record<string, FieldEvaluationResult>) {
        this.results = results;
        this.getApplicableResults.bind(this);
        this.evaluationTimestamp = new Date();
    }
    /**
     * Flatten contents into single list of {@link RuleEvaluationResult}s
     */
    getAllRuleResults(): RuleEvaluationResult[] {
        return Object.values(this.results)
            .map(x => x.ruleResults)
            .reduce(flat, []);
    }
    getFieldResults(): Record<string, FieldEvaluationResult> {
        return this.results;
    }

    getApplicableResults(): RuleEvaluationResult[] {
        return this.getAllRuleResults()
            .filter(r => ConditionEvaluationResult.isApplicable(r.conditionEvaluationResult));
    }

    static empty(): EntryPointResult {
        return new EntryPointResult({});
    }
}
