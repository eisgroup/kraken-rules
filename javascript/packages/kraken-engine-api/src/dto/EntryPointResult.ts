/*
 *  Copyright 2022 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { FieldEvaluationResult } from './FieldEvaluationResult'
import { RuleEvaluationResults } from './RuleEvaluationResults'
import RuleEvaluationResult = RuleEvaluationResults.RuleEvaluationResult

/**
 * Contains evaluation results from specific entry point
 */
export interface EntryPointResult {
    readonly evaluationTimestamp: Date
    readonly ruleTimeZoneId: string

    /**
     * Returns entry point result mapped by field name.
     */
    getFieldResults(): Record<string, FieldEvaluationResult>

    /**
     * Flattens and returns field results into an array of {@link RuleEvaluationResult}
     */
    getAllRuleResults(): RuleEvaluationResult[]

    /**
     * Flattens and returns field results into an array of {@link RuleEvaluationResult}
     * filtered by rule condition evaluation status.
     */
    getApplicableResults(): RuleEvaluationResult[]
}
