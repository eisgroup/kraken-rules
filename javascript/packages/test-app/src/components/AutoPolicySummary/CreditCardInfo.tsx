/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import * as React from "react";
import { Row } from "antd";
import { Moment } from "moment";
import { optional } from "declarative-js";

import { InnerInputsComponentProps, SingleField, ManyFields } from "../core/field/SingleField";
import { trans } from "../core/DateUtils";
import { renderers } from "../core/RenderInputFunctions";

import { BillingAddress } from "./BillingAddres";

import { TestProduct } from "kraken-test-product";
import domain = TestProduct.kraken.testproduct.domain;
import { withMetadata } from "../core/ContextHOC";
import { ContextDefinitionInfo } from "../core/field/ContextDefinitionInfo";

class Component extends React.Component<InnerInputsComponentProps<domain.CreditCardInfo>> {
    onCardTypeChange = (event: React.FormEvent<HTMLInputElement>) => {
        this.props.onChange(Object.assign({}, this.props.value, { cardType: event.currentTarget.value }));
    }

    onCardNumberChange = (event: React.FormEvent<HTMLInputElement>) => {
        this.props.onChange(Object.assign({}, this.props.value, { cardNumber: event.currentTarget.value }));
    }

    onCvvChange = (cvv: React.ReactText) => {
        this.props.onChange(Object.assign({}, this.props.value, { cvv }));
    }

    onExpirationDateChange = (e: Moment) => {
        this.props.onChange(Object.assign({}, this.props.value, { expirationDate: trans.toDate(e) }));
    }

    onCreditCardLimitAmountChange = (e: React.ReactText) => {
        this.props.onChange(Object.assign(
            {},
            this.props.value,
            { cardCreditLimitAmount: { amount: Number(e), currency: "USD" } }
        ));
    }

    onBillingAddressChange = (billingAddress: domain.BillingAddress) => {
        this.props.onChange(Object.assign({}, this.props.value, { billingAddress }));
    }

    onRefToBankChange = (refsToBank: string[]) => {
        this.props.onChange(Object.assign({}, this.props.value, { refsToBank }));
    }

    render(): JSX.Element {
        return (
            <div>
                <Row>
                    <SingleField
                        id={this.props.id}
                        value={this.props.value.cardType}
                        contextName="CreditCardInfo"
                        modelFieldName="cardType"
                        onChange={this.onCardTypeChange}

                        renderInput={renderers.input}
                    />
                    <SingleField
                        id={this.props.id}
                        value={this.props.value.cardNumber}
                        contextName="CreditCardInfo"
                        modelFieldName="cardNumber"
                        onChange={this.onCardNumberChange}

                        renderInput={renderers.input}
                    />
                    <SingleField
                        id={this.props.id}
                        value={this.props.value.cvv}
                        contextName="CreditCardInfo"
                        modelFieldName="cvv"
                        onChange={this.onCvvChange}

                        renderInput={renderers.inputNumber}
                    />
                    <SingleField
                        id={this.props.id}
                        value={trans.toMoment(this.props.value.expirationDate)}
                        contextName="CreditCardInfo"
                        modelFieldName="expirationDate"
                        onChange={this.onExpirationDateChange}

                        renderInput={renderers.date}
                    />
                    <SingleField
                        id={this.props.id}
                        value={optional(this.props.value.cardCreditLimitAmount).map(x => x.amount).orElse(undefined)}
                        contextName="CreditCardInfo"
                        modelFieldName="cardCreditLimitAmount"
                        onChange={this.onCreditCardLimitAmountChange}

                        renderInput={renderers.inputCurrency}
                    />
                    <ManyFields
                        id={this.props.id}
                        value={optional(this.props.value.refsToBank).orElse([])}
                        contextName="CreditCardInfo"
                        modelFieldName="refsToBank"
                        onChange={this.onRefToBankChange}

                        renderInput={renderers.input}
                    />
                </Row>

                <ContextDefinitionInfo contextName="BillingAddress" />
                <BillingAddress
                    id={this.props.value.billingAddress.id}
                    value={this.props.value.billingAddress}
                    onChange={this.onBillingAddressChange}
                />
            </div>
        );
    }
}
export const CreditCardInfo = withMetadata(Component);
