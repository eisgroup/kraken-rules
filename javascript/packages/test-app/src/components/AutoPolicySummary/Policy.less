/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

body {
    margin-top: 0px;
}

.input-width {
    width: 90% !important;
    margin-bottom: 15px !important;
    margin-top: 3px !important;
}

.entry-point-name-select {
    margin: 0 10px 0 0;
}

#buttonPanel {
    width: 100%;
    position: fixed;
    top: 0px;
    background: #1890ffe8;
    padding: 16px;
    right: 0px;
    z-index: 1;
    text-align: right;
    box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.26);
}

#buttonPanel button {
    margin-right: 10px;
}

#buttonPanel div {
    margin-right: 10px;
}

.ContextDefinitionInfo {
    border-radius: 3px;
    padding: 11px;
    background: #90caff59;
    border: 1px solid #2b98fd5e;
    margin: 10px 0px;
}

.sub-entity-box {
    border: 3px solid #80808087;
    padding: 18px;
    margin: 18px;
    border-radius: 3px;
    box-shadow: 0px 0px 20px -3px rgba(0, 0, 0, 0.5);
    background: url('https://www.toptal.com/designers/subtlepatterns/patterns/dot-grid.png');
}
.sub-entity-box > h2 {
    color: rgb(45, 154, 255);
    text-align: center;
    text-transform: capitalize;
}
