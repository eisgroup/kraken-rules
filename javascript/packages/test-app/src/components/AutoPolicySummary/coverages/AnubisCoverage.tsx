/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import * as React from 'react'
import { Row } from 'antd'

import { InnerInputsComponentProps, SingleField } from '../../core/field/SingleField'
import { renderers } from '../../core/RenderInputFunctions'

import { TestProduct } from 'kraken-test-product'
import domain = TestProduct.kraken.testproduct.domain
import { withMetadata } from '../../core/ContextHOC'
import { Moment } from 'moment'
import { trans } from '../../core/DateUtils'

class Component extends React.Component<InnerInputsComponentProps<domain.AnubisCoverage>> {
    onCodeChange = (event: React.FormEvent<HTMLInputElement>) => {
        this.props.onChange(Object.assign({}, this.props.value, { code: event.currentTarget.value }))
    }

    onLimitAmountChange = (limitAmount: React.ReactText) => {
        this.props.onChange(Object.assign({}, this.props.value, { limitAmount }))
    }

    onDeductibleAmountChange = (deductibleAmount: React.ReactText) => {
        this.props.onChange(Object.assign({}, this.props.value, { deductibleAmount }))
    }

    onCultChange = (name: React.FormEvent<HTMLInputElement>) => {
        const cult = { ...this.props.value.cult }
        cult.name = name.currentTarget.value
        this.props.onChange(Object.assign({}, this.props.value, { cult }))
    }

    onCultDateChange = (e: Moment) => {
        const cult = { ...this.props.value.cult }
        cult.date = trans.toDate(e)
        this.props.onChange(Object.assign({}, this.props.value, { cult }))
    }

    render(): JSX.Element {
        return (
            <Row>
                <SingleField
                    id={this.props.id}
                    value={this.props.value.code}
                    contextName='AnubisCoverage'
                    modelFieldName='code'
                    onChange={this.onCodeChange}
                    renderInput={renderers.input}
                />
                <SingleField
                    id={this.props.id}
                    value={this.props.value.limitAmount}
                    contextName='AnubisCoverage'
                    modelFieldName='limitAmount'
                    onChange={this.onLimitAmountChange}
                    renderInput={renderers.inputNumber}
                />
                <SingleField
                    id={this.props.id}
                    value={this.props.value.deductibleAmount}
                    contextName='AnubisCoverage'
                    modelFieldName='deductibleAmount'
                    onChange={this.onDeductibleAmountChange}
                    renderInput={renderers.inputNumber}
                />
                <SingleField
                    id={this.props.id}
                    value={this.props.value.cult.name}
                    contextName='AnubisCoverage'
                    modelFieldName='cultName'
                    onChange={this.onCultChange}
                    renderInput={renderers.input}
                />
                <SingleField
                    id={this.props.id}
                    value={trans.toMoment(this.props.value.cult.date)}
                    contextName='AnubisCoverage'
                    modelFieldName='cultDate'
                    onChange={this.onCultDateChange}
                    renderInput={renderers.date}
                />
            </Row>
        )
    }
}

export const AnubisCoverage = withMetadata(Component)
