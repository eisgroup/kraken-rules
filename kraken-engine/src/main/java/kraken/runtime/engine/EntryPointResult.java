package kraken.runtime.engine;

import kraken.annotations.API;
import kraken.runtime.engine.dto.FieldEvaluationResult;
import kraken.runtime.engine.dto.RuleEvaluationResult;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Contains evaluation results from specific entry point
 *
 * @author rimas
 * @since 1.0
 */
@API
public class EntryPointResult {

    /**
     * Field evaluation results, mapped to field id string
     */
    private Map<String, FieldEvaluationResult> fieldResults;

    private transient LocalDateTime evaluationTimeStamp;
    /**
     * Default constructor
     */
    public EntryPointResult() {
        fieldResults = new HashMap<>();
        evaluationTimeStamp = LocalDateTime.now();
    }

    public EntryPointResult(Map<String, FieldEvaluationResult> fieldResults, LocalDateTime evaluationTimeStamp) {
        this.fieldResults = new HashMap<>(fieldResults);
        this.evaluationTimeStamp = evaluationTimeStamp;
    }

    /**
     * Flatten contents into single list of {@link RuleEvaluationResult}s
     *
     * @return  list of all contained {@link RuleEvaluationResult}s
     */
    public List<RuleEvaluationResult> getAllRuleResults() {
        return fieldResults.entrySet().stream()
                .flatMap(e -> e.getValue().getRuleResults().stream())
                .collect(Collectors.toList());
    }

    public List<RuleEvaluationResult> getApplicableRuleResults() {
        return fieldResults.entrySet().stream()
                .flatMap(e -> e.getValue().getRuleResults().stream())
                .filter(r -> r.getConditionEvaluationResult().isApplicable())
                .collect(Collectors.toList());
    }

    public LocalDateTime getEvaluationTimeStamp() {
        return evaluationTimeStamp;
    }

    public Map<String, FieldEvaluationResult> getFieldResults() {
        return fieldResults;
    }

}
