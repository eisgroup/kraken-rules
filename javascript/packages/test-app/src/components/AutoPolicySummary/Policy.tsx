/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import * as React from "react";

import "antd/lib/radio/style";
import "antd/lib/input/style";
import "antd/lib/input-number/style";
import "antd/lib/date-picker/style";
import "antd/lib/button/style";
import "antd/lib/spin/style";
import "antd/lib/select/style";
import "antd/lib/collapse/style";
import "antd/lib/row/style";
import "antd/lib/list/style";
import "antd/lib/notification/style";
import "antd/lib/slider/style";

import { Button, Spin, Row, Select, Collapse, Popover, Radio } from "antd";

import { SingleField, ManyFields } from "../core/field/SingleField";
import { renderers } from "../core/RenderInputFunctions";
import { entryPointNames } from "../../common/constants";

import { AutoPolicyViewState, EvaluationResponse } from "./Policy.container";
import { TransactionDetails } from "./TransactionDetails";
import { PolicyDetails } from "./PolicyDetails";
import { TermDetails } from "./TermDetails";
import { AccessTrackInfo } from "./AccessTrackInfo";
import { CreditCardInfo } from "./CreditCardInfo";
import { Insured } from "./Insured";
import { Vehicle } from "./Vehicle";
import { Party } from "./Party";
import { UIResultModal } from "./modals/UIResultModal";
import { BackendResultModal } from "./modals/BackendResultModal";
import { ModelModal } from "./modals/ModelModal";
import { ImportModal } from "./modals/ImportModal";

import { TestProduct } from "kraken-test-product";
import domain = TestProduct.kraken.testproduct.domain;
import { ErrorBoundary } from "../../ErrorBoundary";
import { ContextDefinitionInfo } from "../core/field/ContextDefinitionInfo";
import { EntityBox } from "../core/EntityBox";
import { Referer } from "./Referer";
import { ExpressionTreeModal } from "./modals/ExpressionTreeModal";

export interface AutoPolicyCallbackFunctionProps {
    onClickShowExpressionModal: () => void;
    onClickShowUIResultModal: () => void;
    onClickShowBackendResultModal: () => void;
    onClickShowModelModal: () => void;
    onClickShowImportModal: () => void;
    onImportableModelSubmit: () => void;
    onImportableModelChange: (e: React.FormEvent<HTMLTextAreaElement>) => void;
    onValidationEntryPointChange: (validationEntryPointName: string) => void;
    validate: () => void;
    onClickExportCurrentState: () => void;
    onClickResetUI: () => void;
    onStateChange: (state: string) => void;
    onPolicyChange: (e: React.FormEvent<HTMLInputElement>) => void;
    onCreatedFromPolicyRevChange: (e: React.ReactText) => void;
    onPoliciesChange: (e: string[]) => void;
    onTransDetailsChange: (det: domain.TransactionDetails) => void;
    onPolicyDetailsChange: (det: domain.PolicyDetail) => void;
    onTermDetailsChange: (det: domain.TermDetails) => void;
    onAccessTrackInfoChange: (accessTrackInfo: domain.AccessTrackInfo) => void;
    onCreditCardInfoChange: (creditCardInfo: domain.CreditCardInfo) => void;
    onInsuredChange: (insured: domain.Insured) => void;
    onVehicleChange: (changedVehicleIndex: Number) => (det: domain.Vehicle) => void;
    onPartyChange: (party: domain.Party) => void;
    onRefToCustomerChange: (e: React.FormEvent<HTMLInputElement>) => void;
    onClickResetEntryPoint: () => void;
    onRefererChange: (r: domain.Referer) => void;
}

export interface AutoPolicyProps extends AutoPolicyCallbackFunctionProps {
    autoPolicyViewState: AutoPolicyViewState & EvaluationResponse;
}

export class Policy extends React.Component<AutoPolicyProps> {
    render(): JSX.Element {
        const { autoPolicyViewState } = this.props;
        return (
            <Spin spinning={false} size="large">
                {/* <Spin spinning={autoPolicyViewState.isLoading} size="large"> */}
                <div>
                    <b>{autoPolicyViewState.buildInfo && autoPolicyViewState.buildInfo.date}</b>
                    <UIResultModal
                        onClickShowModal={this.props.onClickShowUIResultModal}
                        modalVisibility={autoPolicyViewState.isUIResultModalVisible}
                        fieldResults={autoPolicyViewState.fieldResults}
                        allRuleResults={autoPolicyViewState.allRuleResults}
                        validationStatus={autoPolicyViewState.validationStatus}
                    />
                    <BackendResultModal
                        onClickShowModal={this.props.onClickShowBackendResultModal}
                        modalVisibility={autoPolicyViewState.isBackendResultModalVisible}
                        backendRawResult={autoPolicyViewState.backendRawResult}
                        backendDefaultResult={autoPolicyViewState.backendDefaultResult}
                        backendValidationResult={autoPolicyViewState.backendValidationResult}
                    />
                    <ModelModal
                        onClickShowModal={this.props.onClickShowModelModal}
                        modalVisibility={autoPolicyViewState.isModelModalVisible}
                        modelBeforeEvaluation={autoPolicyViewState.modelBeforeEvaluation}
                        modelAfterEvaluation={autoPolicyViewState.modelAfterEvaluation}
                        modelCurrentState={autoPolicyViewState.modelCurrentState}
                    />
                    <ExpressionTreeModal
                        modalVisibility={autoPolicyViewState.isExpressionModalVisible}
                        onClickShowModal={this.props.onClickShowExpressionModal}
                    />
                    <ImportModal
                        modalVisibility={autoPolicyViewState.isImportModalVisible}
                        onClickShowModal={this.props.onClickShowImportModal}
                        onImportableModelSubmit={this.props.onImportableModelSubmit}
                        onImportableModelChange={this.props.onImportableModelChange}
                        importableModelState={autoPolicyViewState.importableModelState}
                    />
                    <div id="buttonPanel">
                        <Select
                            className="entry-point-name-select"
                            defaultValue={entryPointNames.UI}
                            onChange={this.props.onValidationEntryPointChange}
                            style={{ width: 200 }}
                        >
                            {Object.keys(entryPointNames).map((entryPointName: string) => (
                                <Select.Option key={entryPointName} value={entryPointName}>
                                    {entryPointName}
                                </Select.Option>
                            ))}
                        </Select>
                        <Button
                            loading={this.props.autoPolicyViewState.isLoading}
                            type="dashed"
                            icon="code-o"
                            onClick={this.props.validate}
                        >
                            Validate
                        </Button>
                        <Button type="dashed" icon="eye-o" onClick={this.props.onClickShowUIResultModal}>
                            UI result
                        </Button>
                        <Button
                            type={this.props.autoPolicyViewState.isBackendFailed ? "danger" : "dashed"}
                            icon="eye-o"
                            onClick={this.props.onClickShowBackendResultModal}
                        >
                            Backend result
                        </Button>
                        <Button type="dashed" icon="eye-o" onClick={this.props.onClickShowModelModal}>
                            Model
                        </Button>
                        <Button type="dashed" icon="experiment" onClick={this.props.onClickShowExpressionModal}>
                            Expression
                        </Button>
                        <Button type="dashed" icon="upload" onClick={this.props.onClickShowImportModal}>
                            Import model
                        </Button>
                        <Button type="dashed" icon="download" onClick={this.props.onClickExportCurrentState}>
                            Export state
                        </Button>
                        <Button type="danger" icon="delete" onClick={this.props.onClickResetUI}>
                            Reset form
                        </Button>
                        <Button type="danger" icon="reload" onClick={this.props.onClickResetEntryPoint}>
                            Reset EntryPoint
                        </Button>
                        <br />
                        <span>
                            Status <b id="status">{autoPolicyViewState.isLoading ? "Loading" : "OK"}</b>&nbsp;
                        </span>
                        <br />
                    </div>
                    <ErrorBoundary>
                        <EntityBox title="Policy" >
                            <ContextDefinitionInfo contextName="Policy" />
                            <Row>
                                <SingleField
                                    id={autoPolicyViewState.autoPolicy.id}
                                    value={autoPolicyViewState.autoPolicy.state}
                                    contextName="Policy"
                                    modelFieldName="state"
                                    onChange={this.props.onStateChange}
                                    // tslint:disable-next-line: max-line-length
                                    info="dimension: state, use 'Dimensional' entrypoint to get rules based on that dimension"
                                    renderInput={renderers.select}
                                    selections={["CA", "AZ"]}

                                />
                                <SingleField
                                    id={autoPolicyViewState.autoPolicy.id}
                                    value={autoPolicyViewState.autoPolicy.policyNumber}
                                    contextName="Policy"
                                    modelFieldName="policyNumber"
                                    onChange={this.props.onPolicyChange}
                                    renderInput={renderers.input}
                                />
                                <SingleField
                                    id={autoPolicyViewState.autoPolicy.id}
                                    value={autoPolicyViewState.autoPolicy.createdFromPolicyRev}
                                    contextName="Policy"
                                    modelFieldName="createdFromPolicyRev"
                                    onChange={this.props.onCreatedFromPolicyRevChange}
                                    renderInput={renderers.inputNumber}
                                />
                                <SingleField
                                    id={autoPolicyViewState.autoPolicy.id}
                                    value={autoPolicyViewState.autoPolicy.refToCustomer}
                                    contextName="Policy"
                                    modelFieldName="refToCustomer"
                                    onChange={this.props.onRefToCustomerChange}
                                    renderInput={renderers.inputDisabled}
                                    info="This field is readonly"
                                />
                                <ManyFields
                                    id={autoPolicyViewState.autoPolicy.id}
                                    value={autoPolicyViewState.autoPolicy.policies}
                                    contextName="Policy"
                                    modelFieldName="policies"
                                    onChange={this.props.onPoliciesChange}
                                    renderInput={renderers.input}
                                />
                            </Row>
                            <TransactionDetails
                                id={autoPolicyViewState.autoPolicy.id}
                                value={autoPolicyViewState.autoPolicy.transactionDetails}
                                onChange={this.props.onTransDetailsChange}
                            />
                            <PolicyDetails
                                id={autoPolicyViewState.autoPolicy.id}
                                value={autoPolicyViewState.autoPolicy.policyDetail}
                                onChange={this.props.onPolicyDetailsChange}
                            />
                            <TermDetails
                                id={autoPolicyViewState.autoPolicy.id}
                                value={autoPolicyViewState.autoPolicy.termDetails}
                                onChange={this.props.onTermDetailsChange}
                            />
                            <AccessTrackInfo
                                id={autoPolicyViewState.autoPolicy.id}
                                value={autoPolicyViewState.autoPolicy.accessTrackInfo}
                                onChange={this.props.onAccessTrackInfoChange}
                            />
                        </EntityBox>
                        <EntityBox title="Referer" >
                            <ContextDefinitionInfo contextName="Referer" />
                            <Referer
                                id={autoPolicyViewState.autoPolicy.referer.id}
                                value={autoPolicyViewState.autoPolicy.referer}
                                onChange={this.props.onRefererChange}
                            />
                        </EntityBox>
                        <EntityBox title="CreditCardInfo" >
                            <ContextDefinitionInfo contextName="CreditCardInfo" />
                            <CreditCardInfo
                                id={autoPolicyViewState.autoPolicy.billingInfo.creditCardInfo.id}
                                value={autoPolicyViewState.autoPolicy.billingInfo.creditCardInfo}
                                onChange={this.props.onCreditCardInfoChange}
                            />
                        </EntityBox>
                        <EntityBox title="Insured" >
                            <ContextDefinitionInfo contextName="Insured" />
                            <Insured
                                id={autoPolicyViewState.autoPolicy.insured.id}
                                value={autoPolicyViewState.autoPolicy.insured}
                                onChange={this.props.onInsuredChange}
                            />
                        </EntityBox>
                        <EntityBox title="Risk Items" >
                            <Collapse>
                                {autoPolicyViewState.autoPolicy.riskItems.map((riskItem, index) => (
                                    <Collapse.Panel header={"Vehicle " + index} key={riskItem.id}>
                                        <Vehicle
                                            id={autoPolicyViewState.autoPolicy.riskItems[index].id}
                                            value={autoPolicyViewState.autoPolicy.riskItems[index]}
                                            onChange={this.props.onVehicleChange(index)}
                                        />
                                    </Collapse.Panel>
                                ))}
                            </Collapse>
                        </EntityBox>
                        <EntityBox title="Party" >
                            <Party
                                id={autoPolicyViewState.autoPolicy.parties[0].id}
                                value={autoPolicyViewState.autoPolicy.parties[0]}
                                onChange={this.props.onPartyChange}
                            />
                        </EntityBox>
                    </ErrorBoundary>
                </div>
            </Spin >
        );
    }
}
