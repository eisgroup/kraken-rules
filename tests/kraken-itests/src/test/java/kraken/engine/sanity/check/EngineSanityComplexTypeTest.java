/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package kraken.engine.sanity.check;

import com.google.common.collect.ImmutableList;
import kraken.runtime.engine.EntryPointResult;
import kraken.testproduct.domain.AddressInfo;
import kraken.testproduct.domain.AddressLine;
import kraken.testproduct.domain.AddressLine1;
import kraken.testproduct.domain.AddressLine2;
import kraken.testproduct.domain.Policy;
import kraken.testproduct.domain.CreditCardInfo;
import kraken.testproduct.domain.AnubisCoverage;
import kraken.testproduct.domain.Vehicle;
import kraken.utils.MockAutoPolicyBuilder;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import static io.github.jsonSnapshot.SnapshotMatcher.start;
import static io.github.jsonSnapshot.SnapshotMatcher.validateSnapshots;
import static kraken.testing.matchers.KrakenMatchers.*;
import static org.junit.Assert.assertThat;
import static kraken.test.KrakenItestMatchers.matchesSnapshot;

/**
 * @author mulevicius
 */
public class EngineSanityComplexTypeTest extends SanityEngineBaseTest {

    @BeforeClass
    public static void beforeAll() {
        start();
    }

    @AfterClass
    public static void afterAll() {
        validateSnapshots();
    }

    @Test
    public void shouldReferenceComplexCollectionFieldType(){
        AddressInfo addressInfo = new AddressInfo();
        addressInfo.setAddressLine1(new AddressLine1("addressLine1"));
        addressInfo.setAddressLine2(new AddressLine2("addressLine2"));

        Vehicle vehicle = new Vehicle();
        vehicle.setAddressInfo(addressInfo);
        vehicle.setAnubisCoverages(ImmutableList.of(new AnubisCoverage(), new AnubisCoverage()));

        Policy policy = new MockAutoPolicyBuilder()
                .addCreditCardInfo(new CreditCardInfo())
                .addRiskItems(ImmutableList.of(vehicle))
                .build();

        final EntryPointResult result = engine.evaluate(policy, "complex-field-type-test");

        assertThat(result, hasRuleResults(2));
        assertThat(result, hasValidationFailures(1));
        assertThat(result, hasNoIgnoredRules());
        assertThat(result, matchesSnapshot());
    }

}
