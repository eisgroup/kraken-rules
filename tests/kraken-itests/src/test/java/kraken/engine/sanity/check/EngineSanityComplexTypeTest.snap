kraken.engine.sanity.check.EngineSanityComplexTypeTest.shouldReferenceComplexCollectionFieldType=[
  {
    "AddressInfo:-1:addressLines": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "AddressInfo",
        "fieldName": "addressLines",
        "fieldPath": "addressLines"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "ctype-addressLinesMustBeThree",
            "context": "AddressInfo",
            "targetPath": "addressLines",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": false,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "Vehicle:-1:anubisCoverages": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "Vehicle",
        "fieldName": "anubisCoverages",
        "fieldPath": "anubisCoverages"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "ctype-thereMustBeAnubisCoveragesAccessibleByProperty",
            "context": "Vehicle",
            "targetPath": "anubisCoverages",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    }
  }
]